### PR TITLE
Update logistic regression model and refactorization

### PR DESCRIPTION
**Pull request by chanteriam**
## Describe your changes
Updates the Logistic Regression model to run agains the four processed datasets and adds the structure of a class to run a gradient boosted tree model.

## Checklist before requesting a review
- [x] The code runs successfully.

<details><summary>Output from make run</summary>

```
(poetry-python-project-py3.11) michaelp@MacBook-Air-3 supreme-court-ml-predictions % make lint
ruff "supreme_court_predictions"/ test/
(poetry-python-project-py3.11) michaelp@MacBook-Air-3 supreme-court-ml-predictions % make run   
python -m "supreme_court_predictions" --get-data --clean-data --describe-data --process-data --logistic-regression
oh, what up?
Loading Supreme Court Corpus Data...
Dataset already exists at /Users/michaelp/.convokit/downloads/supreme-corpus
Working in /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/convokit/
Data will be saved to /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/clean_convokit/
Parsing cases...
Parsing speakers...
Parsing conversations metadata...
Parsing utterances...
Data saved to /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/clean_convokit/
Data will be saved to: 
/Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/clean_convokit/
Spacy module successfully loaded.
Spacy tokenization complete.
Working in /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/clean_convokit/
Data will be saved to /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/statistics/
Grabbing case descriptive statistics...
Grabbing advocates descriptive statistics...
Grabbing speakers descriptive statistics...
Grabbing voters descriptive statistics...
Grabbing utterances descriptive statistics...
Data saved to /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/statistics/
Working in /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/clean_convokit/
Data will be saved to /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/processed/
Grabbing token aggregation for all cases...
Grabbing token aggregation for advocates...
Grabbing token aggregation for adversaries...
Grabbing token aggregation for judges...
Exporting files...
Data saved to /Users/michaelp/Documents/GitHub/supreme-court-ml-predictions/supreme_court_predictions/data/processed/
------------------------------------------
Running regression on total_utterances...
Creating bag of words
Starting the Logistic Regression
Accuracy score: 0.5405405405405406
------------------------------------------
------------------------------------------
Running regression on judge_utterances...
Creating bag of words
Starting the Logistic Regression
Accuracy score: 0.5205479452054794
------------------------------------------
------------------------------------------
Running regression on advocate_utterances...
Creating bag of words
Starting the Logistic Regression
Accuracy score: 0.7567567567567568
------------------------------------------
------------------------------------------
Running regression on adversary_utterances...
Creating bag of words
Starting the Logistic Regression
Accuracy score: 0.7837837837837838
------------------------------------------
(poetry-python-project-py3.11) michaelp@MacBook-Air-3 supreme-court-ml-predictions % 
```
</details>

- [x] I ran `make lint` and have made the bulk of changes that it has requested.

### CODE REVIEW

File: supreme_court_predictions/models/logistic_regression.py [hunk 1]
```diff
@@ -17,25 +18,38 @@ class LogisticRegression:
...
+        )
+        self.judge_utterances = pd.read_pickle(
+            self.local_path + "judge_aggregations.p"
+        )
+
+        self.run_regression()
 
-    def logistic_regression(self):
+    @staticmethod
+    def logistic_regression(df):
```
Comment (michplunkett) [inline supreme_court_predictions/models/logistic_regression.py:39]: We don't use `self` here, so I made it static.

**Review by JJModern — APPROVED** [review]
looks good

**Comment by michplunkett** [comment]
@JJModern Can you add the docstrings to the file and functions on this branch?

File: supreme_court_predictions/models/logistic_regression.py [hunk 2]
```diff
@@ -1,48 +1,101 @@
...
-        bag_of_words_y = np.random.randint(0, 2, len(vectorize_document))
+        bag_of_words_y = df.loc[:, "win_side"]
 
         X_train, X_test, y_train, y_test = train_test_split(
-            bag_of_words_x, bag_of_words_y, test_size=0.25, random_state=123
+            bag_of_words_x,
+            bag_of_words_y,
+            test_size=0.20,
+            random_state=123,
+            stratify=bag_of_words_y,
```
Comment (JJModern) [inline supreme_court_predictions/models/logistic_regression.py:65]: try catch block and stratify allow smaller datsets to be used. Now the actual LR is done.

File: pyproject.toml [hunk 3]
```diff
@@ -19,6 +19,7 @@ openpyxl = "^3.1.2"
 scikit-learn = "^1.2.2"
 spacy = "^3.5.2"
 ruff = "^0.0.263"
+xgboost = "^1.7.5"
```
Comment (michplunkett) [inline pyproject.toml:22]: `xgboost` is added for the next batch of models.

File: supreme_court_predictions/models/logistic_regression.py [hunk 4]
```diff
@@ -1,48 +1,97 @@
...
+        ]
+
+        for df_name, df in dfs:
+            try:
+                print("------------------------------------------")
+                print(f"Running regression on {df_name}...")
+                acc = self.create(df)
+                print(f"Accuracy score: {acc}")
+                print("------------------------------------------")
+            except ValueError:
```
Comment (michplunkett) [inline supreme_court_predictions/models/logistic_regression.py:94]: The linter complained if it did not have an error assigned so I went with `ValueError` based off some searching around.

File: supreme_court_predictions/models/model.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+"""
+This file contains the class that is the basis for all models in this package.
+"""
+from abc import ABC, abstractmethod
+
+
+class Model(ABC):
```
Comment (michplunkett) [inline supreme_court_predictions/models/model.py:7]: Abstract class for all models.

File: supreme_court_predictions/models/xg_boost.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
+"""
```
Comment (michplunkett) [inline supreme_court_predictions/models/xg_boost.py:1]: WIP.

File: supreme_court_predictions/processing/datacleaner.py [hunk 7]
```diff
@@ -23,16 +23,13 @@ class DataCleaner:
...
-    def __init__(self, downloaded_corpus, save_data=True):
+    def __init__(self):
         self.cases_df = None
         self.clean_case_ids = None  # stores the case IDs to use
         self.clean_utterances_list = None
         self.speakers_df = None
         self.utterances_df = None
 
-        self.downloaded_corpus = downloaded_corpus
-        self.save_data = save_data
```
Comment (michplunkett) [inline supreme_court_predictions/processing/datacleaner.py:34]: I can't think of a reason to not save the data. This step is also called after `get-data`, so the corpus flag is redundant.

**Review by michplunkett — APPROVED** [review]

**Review by FedericoDM — APPROVED** [review]
Looks good, next step is to get XGBoost, Random Forests and some ensemble up and running.
